### PR TITLE
Optimized Infer Request Scheduling

### PR DESCRIPTION
**Pull request by apankratovantonp**
Less operations with shared queues.

### CODE REVIEW

File: inference-engine/src/multi_device/multi_device_exec_network.hpp [hunk 1]
```diff
@@ -56,15 +57,40 @@ class ThreadSafeQueue {
...
+    std::queue<T>   _queue;
+    std::mutex      _mutex;
+};
+template <typename T>
+class ThreadSafeBoundedQueue {
+public:
+    ThreadSafeBoundedQueue() = default;
+    bool try_push(T value) {
+        std::lock_guard<std::mutex> lock(_mutex);
+        if (_capacity) {
```
Comment (myshevts) [inline inference-engine/src/multi_device/multi_device_exec_network.hpp:70]: Shall you check that the capacity is >= than current size+1?

File: inference-engine/src/multi_device/multi_device_exec_network.hpp [hunk 2]
```diff
@@ -56,15 +57,40 @@ class ThreadSafeQueue {
...
+            _queue.pop();
+            return true;
+        } else {
+            return false;
+        }
+    }
+    void set_capacity(std::size_t newCapacity) {
         std::lock_guard<std::mutex> lock(_mutex);
-        return _queue.empty();
+        _capacity = newCapacity;
```
Comment (ilya-lavrenov) [inline inference-engine/src/multi_device/multi_device_exec_network.hpp:87]: `_capacity` is bool, `newCapacity` is size_t. Is it expected?

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 3]
```diff
@@ -92,52 +93,59 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const DeviceMap<Infer
...
-            }
+        idleWorkerRequests = &(_idleWorkerRequests[device.deviceName]);
+        if (idleWorkerRequests->try_pop(workerRequestPtr)) {
+            break;
         }
     }
+    if (workerRequestPtr != nullptr) {
+        IdleGuard idleGuard{workerRequestPtr, *idleWorkerRequests};
+        _thisWorkerInferRequest = workerRequestPtr;
+        auto captchuredTask = std::move(inferPipelineTask);
```
Comment (ilya-lavrenov) [inline inference-engine/src/multi_device/multi_device_exec_network.cpp:122]: `captchuredTask` -> `capturedTask`

File: inference-engine/src/multi_device/multi_device_exec_network.hpp (hunk 2 above)
Comment (apankratovantonp) [reply inference-engine/src/multi_device/multi_device_exec_network.hpp:87]: Yes in this model we need just two states of queue enabled/disabled.

File: inference-engine/src/multi_device/multi_device_exec_network.hpp (hunk 1 above)
Comment (apankratovantonp) [reply inference-engine/src/multi_device/multi_device_exec_network.hpp:70]: No

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 4]
```diff
@@ -92,52 +93,61 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const DeviceMap<Infer
...
         std::lock_guard<std::mutex> lock(_mutex);
         _devicePriorities.clear();
     }
-    _terminate = true;
     /* NOTE: The only threads that use `MultiDeviceExecutableNetwork` Context are those that are used by Worker infer requests.
      *       But AsyncInferRequest destructor should waits for all asynchronous tasks that are used by the request
      */
+    auto devices = [&] {
+        std::lock_guard<std::mutex> lock(_mutex);
+        return _devicePriorities;
```
Comment (myshevts) [inline inference-engine/src/multi_device/multi_device_exec_network.cpp:146]: _device_priorities is not used, or is it?

File: inference-engine/src/multi_device/multi_device_exec_network.cpp [hunk 5]
```diff
@@ -80,10 +80,11 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const DeviceMap<Infer
         auto& idleWorkerRequests = _idleWorkerRequests[device];
         workerRequests.resize(numRequests);
         auto* idleWorkerRequestsPtr = &(idleWorkerRequests);
+        idleWorkerRequests.set_capacity(numRequests);
         for (auto&& workerRequest : workerRequests) {
             workerRequest._inferRequest = network.CreateInferRequest();
             auto* workerRequestPtr = &workerRequest;
-            idleWorkerRequests.push(workerRequestPtr);
+            IE_ASSERT(idleWorkerRequests.try_push(workerRequestPtr) == true);
```
Comment (myshevts) [inline inference-engine/src/multi_device/multi_device_exec_network.cpp:87]: why not 'push' and no assert? (as it was initially)

File: inference-engine/src/multi_device/multi_device_exec_network.cpp (hunk 4 above)
Comment (myshevts) [reply inference-engine/src/multi_device/multi_device_exec_network.cpp]: fixed

**Comment by myshevts** [comment]
@apankratovantonp , I made further brushing to this PR (esp last commit) pls look

**Review by ilya-lavrenov — APPROVED** [review]

**Comment by ilya-lavrenov** [comment]
@apankratovantonp please, merge with latest master to make CI green

**Pull request merged**